src/common/auth.rs

```diff
@@ -45,33 +55,68 @@ impl AuthKeys {
     /// - `Ok(None)` if the request is allowed through the API key.
     /// - `Ok(Some(claims))` if the request is allowed through the JWT token.
     /// - `Err(description)` if the request is not allowed.
-    pub fn validate_request<'a>(
+    pub async fn validate_request<'a>(
         &self,
         get_header: impl Fn(&'a str) -> Option<&'a str>,
         is_read_only: bool,
-    ) -> Result<Option<Claims>, &'static str> {
+    ) -> Result<Option<Claims>, String> {
         let Some(key) = get_header("api-key")
             .or_else(|| get_header("authorization").and_then(|v| v.strip_prefix("Bearer ")))
         else {
-            return Err("Must provide an API key or an Authorization bearer token");
+            return Err("Must provide an API key or an Authorization bearer token".to_string());
         };
 
         if self.can_write(key) || (is_read_only && self.can_read(key)) {
             return Ok(None);
         }
 
         if !is_read_only && self.can_read(key) {
-            return Err("Write access denied");
+            return Err("Write access denied".to_string());
         }
 
         if let Some(claims) = self.jwt_parser.as_ref().and_then(|p| p.decode(key).ok()) {
-            if !claims.w.unwrap_or(false) && !is_read_only {
-                return Err("Write access denied");
+            let Claims {
+                exp: _, // already validated on decoding
+                w: write_access,
+                value_exists,
+                payload: _,     //
+                collections: _, // will be validated in TableOfContent
+            } = &claims;
+
+            if !write_access.unwrap_or(false) && !is_read_only {
+                return Err("Write access denied".to_string());
             }
+
+            if let Some(value_exists) = value_exists {
+                let scroll_req = ScrollRequestInternal {
+                    offset: None,
+                    limit: Some(1),
+                    filter: Some(value_exists.as_filter()),
+                    with_payload: Some(WithPayloadInterface::Bool(false)),
+                    with_vector: WithVector::Bool(false),
+                    order_by: None,
+                };
+                let res = self
+                    .toc
+                    .scroll(
+                        &value_exists.collection,
+                        scroll_req,
+                        None,
+                        ShardSelectorInternal::All,
+                        None,
+                    )
+                    .await
```

Comment:
I'd prefer to extract this code to a separate function to keep `validate_request()` simple.